fold/fold.go

```diff
@@ -0,0 +1,895 @@
+// Mathews, Sabina, Zuker and Turner, 1999
+// https://www.ncbi.nlm.nih.gov/pubmed/10329189
+// Args:
+//
+//	seq: The sequence to Fold
+//	temp: The temperature the Fold takes place in, in Celsius
+//
+// Returns a slice of NucleicAcidStructure with the energy and description,
+// i.e. stacks, bulges, hairpins, etc.
+func Fold(seq string, temp float64) ([]NucleicAcidStructure, error) {
+	foldContext, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		return nil, fmt.Errorf("error creating folding context: %w", err)
+	}
+
+	// get the minimum free energy structure out of the cache
+	return Traceback(0, len(seq)-1, foldContext), nil
+}
+
+// MinimumFreeEnergy folds the sequence and return just the delta G of the structure
+// Args:
+//
+//	seq: The sequence to fold
+//	temp: The temperature to fold at
+//
+// Returns: the minimum free energy of the folded sequence
+func MinimumFreeEnergy(seq string, temp float64) (float64, error) {
+	NucleicAcidStructures, err := Fold(seq, temp)
+	if err != nil {
+		return 0, fmt.Errorf("error folding: %w", err)
+	}
+	summedEnergy := 0.0
+	for _, structure := range NucleicAcidStructures {
+		summedEnergy += structure.Energy
+	}
+	return RoundFloat(summedEnergy, 2), nil
+}
+
+// Get the dot bracket notation for a secondary structure.
+// Args:
+//
+//	NucleicAcidStructures: A list of NucleicAcidStructure, usually from the fold function
+//
+// # Returns the dot-bracket notation of the secondary structure
+//
+// Dot-bracket notation, consisting in a balanced parentheses string composed
+// by a three-character alphabet {.,(,)}, that can be unambiguously converted
+// in the RNA secondary structure. See example_test.go for a small example.
+func DotBracket(NucleicAcidStructures []NucleicAcidStructure) string {
+	maxj := 0
+	for _, structure := range NucleicAcidStructures {
+		for _, innerSubsequence := range structure.Inner {
+			if innerSubsequence.End > maxj {
+				maxj = innerSubsequence.End
+			}
+		}
+	}
+	maxj += 1
+	result := make([]byte, maxj)
+	for i := range result {
+		result[i] = '.'
+	}
+	for _, structure := range NucleicAcidStructures {
+		if len(structure.Inner) == 1 {
+			innerSubsequence := structure.Inner[0]
+			result[innerSubsequence.Start] = '('
+			result[innerSubsequence.End] = ')'
+		}
+	}
+	return string(result)
+}
+
+// Find and return the lowest free energy structure in the subsequence starting
+// at start and terminating at end.
+//
+// Figure 2B in Zuker and Stiegler, 1981
+// Args:
+//
+//		seq: The sequence being folded
+//		start: The start index
+//		end: The end index (inclusive)
+//	 foldContext: The FoldContext for this sequence
+//
+// Returns the free energy for the subsequence from start to end
+func W(start, end int, foldContext FoldContext) (NucleicAcidStructure, error) {
+	if !foldContext.W[start][end].Equal(DefaultStructure) {
+		return foldContext.W[start][end], nil
+	}
+
+	if end-start < 4 {
```

Comment:
why less than four. What is this magic number?

Reply:
ah. Looks like the original 1981 Zuker paper relies on "pentanucleotides". Off by one?

Reply:
Not sure we are `0` based

Reply:
No, it makes not sense the `0` based I think you are right and it should be `<= 4`